### PR TITLE
feat: Elixir support

### LINKED ISSUE #3212: Add "hex" manager (Elixir/Erlang)
Add support of parsing/ extract hax packages.
Implement hex manages.

### PR DESCRIPTION
**Pull request by zharinov**
Thanks @KonstantinKurganov for your work on Elixir support (#3248), here I'll try to fix the minor issues and merge it.

Closes #3212

### CODE REVIEW

**Comment by zharinov** [comment]
Example repos:
https://github.com/zharinov/swarm/pulls
https://github.com/zharinov/telemetry_metrics/pulls

File: lib/manager/mix/artifacts.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  // istanbul ignore if
+  if (config.binarySource === 'docker') {
+    logger.info('Running mix via docker');
+    cmd = `docker run --rm `;
+    const volumes = [cwd];
+    cmd += volumes.map(v => `-v ${v}:${v} `).join('');
+    const envVars = [];
+    cmd += envVars.map(e => `-e ${e} `);
+    cmd += `-w ${cwd} `;
+    cmd += `renovate/mix mix`;
```
Comment (rarkins) [inline lib/manager/mix/artifacts.ts]: I need to add this

File: lib/manager/mix/artifacts.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        lockFile: lockFileName,
+        stderr: err.message,
+      },
+    };
+  }
+  const duration = process.hrtime(startTime);
+  const seconds = Math.round(duration[0] + duration[1] / 1e9);
+  logger.info({ seconds, type: 'mix.lock' }, 'Updated lockfile');
+  logger.debug('Returning updated mix.lock');
+  const newCargoLockContent = await fs.readFile(localLockFileName, 'utf8');
```
Comment (rarkins) [inline lib/manager/mix/artifacts.ts]:
```suggestion
  const newMixLockContent = await fs.readFile(localLockFileName, 'utf8');
```

File: lib/manager/mix/artifacts.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  if (existingLockFileContent === newCargoLockContent) {
```
Comment (rarkins) [inline lib/manager/mix/artifacts.ts]:
```suggestion
  if (existingLockFileContent === newMixLockContent) {
```

File: lib/manager/mix/artifacts.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    logger.debug('mix.lock is unchanged');
+    return null;
+  }
+  return [
+    {
+      file: {
+        name: lockFileName,
+        contents: newCargoLockContent,
```
Comment (rarkins) [inline lib/manager/mix/artifacts.ts]:
```suggestion
        contents: newMixLockContent,
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/manager/mix/extract.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import { isValid } from '../../versioning/hex';
+import { logger } from '../../logger';
+import { PackageDependency } from '../common';
+
+export function extractPackageFile(content) {
+  logger.trace('mix.extractPackageFile()');
+  const deps = [];
+  const contentArr = content.split('\n');
+  const depSectionRegExp = /defp\s+deps.*do/g;
+  const depMatchRegExp = /{:(\w+),\s*([^:"]+)?:?\s*"([^"]+)",?\s*(organization: "(.*)")?.*}/gm;
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]: regexs should be moved to module scope.

File: lib/manager/mix/extract.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
+import { isValid } from '../../versioning/hex';
+import { logger } from '../../logger';
+import { PackageDependency } from '../common';
+
+export function extractPackageFile(content) {
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]:
```suggestion
export function extractPackageFile(content: string): PackageFile {
```

File: lib/manager/mix/extract.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import { isValid } from '../../versioning/hex';
+import { logger } from '../../logger';
+import { PackageDependency } from '../common';
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]:
```suggestion
import { PackageDependency, PackageFile  } from '../common';
```

File: lib/manager/mix/extract.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+export function extractPackageFile(content) {
+  logger.trace('mix.extractPackageFile()');
+  const deps = [];
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]:
```suggestion
  const deps: PackageDependency[] = [];
```

File: lib/manager/mix/extract.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+            dep.skipReason = 'unsupported-version';
+          }
+
+          if (dep.datasource !== 'hex') {
+            dep.skipReason = 'non-hex depTypes';
+          }
+
+          // Find dep's line number
+          for (let i = 0; i < contentArr.length; i += 1)
+            if (contentArr[i].includes(`:${depName},`)) dep.lineNumber = i;
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]:
```suggestion
            if (contentArr[i].includes(`:${depName},`)) dep.managerData.lineNumber = i;
```

File: lib/manager/mix/extract.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        depMatch = depMatchRegExp.exec(depBuffer);
+        if (depMatch) {
+          const depName = depMatch[1];
+          const datasource = depMatch[2];
+          const currentValue = depMatch[3];
+          const organization = depMatch[5];
+
+          const dep: PackageDependency = {
+            depName,
+            currentValue,
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]:
```suggestion
            currentValue,
            managerData: {},
```

File: lib/manager/mix/extract.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+  for (let lineNumber = 0; lineNumber < contentArr.length; lineNumber += 1) {
+    if (contentArr[lineNumber].match(depSectionRegExp)) {
+      logger.trace(`Matched dep section on line ${lineNumber}`);
+      let depBuffer = '';
+      do {
+        depBuffer += contentArr[lineNumber] + '\n';
+        lineNumber += 1;
+      } while (!contentArr[lineNumber].includes('end'));
+      let depMatch;
```
Comment (ViceIce) [inline lib/manager/mix/extract.ts]:
```suggestion
      let depMatch: RegExpMatchArray;
```

File: lib/manager/mix/update.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import { logger } from '../../logger';
+import { Upgrade } from '../common';
+
+export function updateDependency(fileContent: string, upgrade: Upgrade) {
```
Comment (ViceIce) [inline lib/manager/mix/update.ts]:
```suggestion
export function updateDependency(
  fileContent: string,
  upgrade: Upgrade
): string | null {
```

File: lib/manager/mix/update.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  logger.debug(`mix.updateDependency: ${upgrade.newValue}`);
+
+  const lines = fileContent.split('\n');
+  const lineToChange = lines[upgrade.lineNumber];
```
Comment (ViceIce) [inline lib/manager/mix/update.ts]:
```suggestion
  const lineToChange = lines[upgrade.managerData.lineNumber];
```

File: lib/manager/mix/update.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  if (!lineToChange.includes(upgrade.depName)) return null;
+
+  newLine = lineToChange.replace(/"(.*?)"/, `"${upgrade.newValue}"`);
+
+  if (newLine === lineToChange) {
+    logger.debug('No changes necessary');
+    return fileContent;
+  }
+
+  lines[upgrade.lineNumber] = newLine;
```
Comment (ViceIce) [inline lib/manager/mix/update.ts]:
```suggestion
  lines[upgrade.managerData.lineNumber] = newLine;
```

File: test/manager/mix/update.spec.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  path.resolve(__dirname, './_fixtures/mix.exs'),
+  'utf-8'
+);
+
+describe('lib/manager/mix/update', () => {
+  describe('updateDependency', () => {
+    it('replaces existing value', () => {
+      const upgrade = {
+        depName: 'postgrex',
+        lineNumber: 18,
```
Comment (ViceIce) [inline test/manager/mix/update.spec.ts]:
```suggestion
        managerData: { lineNumber: 18 },
```

File: test/manager/mix/update.spec.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        newValue: '~> 0.8.2',
+      };
+      const res = updateDependency(sample, upgrade);
+      expect(res).not.toEqual(sample);
+      expect(res.includes(upgrade.newValue)).toBe(true);
+    });
+    it('return the same', () => {
+      const upgrade = {
+        depName: 'postgrex',
+        lineNumber: 18,
```
Comment (ViceIce) [inline test/manager/mix/update.spec.ts]:
```suggestion
        managerData: { lineNumber: 18 },
```

File: test/manager/mix/update.spec.ts [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        newValue: '~> 0.8.1',
+      };
+      const res = updateDependency(sample, upgrade);
+      expect(res).toEqual(sample);
+    });
+    it('returns null if wrong line', () => {
+      const upgrade = {
+        depName: 'postgrex',
+        lineNumber: 19,
```
Comment (ViceIce) [inline test/manager/mix/update.spec.ts]:
```suggestion
        managerData: { lineNumber: 19 },
```

File: test/manager/mix/update.spec.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        newValue: '~> 0.8.2',
+      };
+      const res = updateDependency(sample, upgrade);
+      expect(res).toBeNull();
+    });
+    it('returns null for unsupported depType', () => {
+      const upgrade = {
+        depName: 'cowboy',
+        lineNumber: 19,
```
Comment (ViceIce) [inline test/manager/mix/update.spec.ts]:
```suggestion
        managerData: { lineNumber: 19 },
```

File: test/versioning/hex.spec.ts [hunk 19]
```diff
@@ -1,115 +1,104 @@
-import { api as semver } from '../../lib/versioning/hex';
+import { api } from '../../lib/versioning/hex';
```
Comment (ViceIce) [inline test/versioning/hex.spec.ts]: Why this change?

File: lib/manager/mix/artifacts.ts [hunk 20]
```diff
@@ -0,0 +1,120 @@
+import upath from 'upath';
+import fs from 'fs-extra';
+import { platform } from '../../platform';
+import { exec } from '../../util/exec';
+import { logger } from '../../logger';
+
+export async function getArtifacts(
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
export async function updateArtifacts(
```

File: lib/manager/mix/artifacts.ts [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+  packageFileName,
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
  packageFileName: string,
```

File: lib/manager/mix/artifacts.ts [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+  updatedDeps,
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
  updatedDeps: string[],
```

File: lib/manager/mix/artifacts.ts [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+  newPackageFileContent,
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
  newPackageFileContent: string,
```

File: lib/manager/mix/artifacts.ts [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+  config
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
  config: UpdateArtifactsConfig
```

File: lib/manager/mix/artifacts.ts [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+) {
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
): Promise<UpdateArtifactsResult[] | null> {
```

File: lib/manager/mix/artifacts.ts [hunk 26]
```diff
@@ -0,0 +1,120 @@
+import upath from 'upath';
+import fs from 'fs-extra';
+import { platform } from '../../platform';
+import { exec } from '../../util/exec';
+import { logger } from '../../logger';
```
Comment (ViceIce) [inline lib/manager/mix/artifacts.ts]:
```suggestion
import { logger } from '../../logger';
import { UpdateArtifactsConfig, UpdateArtifactsResult } from '../common';
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/hex/index.ts [hunk 27]
```diff
@@ -19,15 +19,27 @@ interface HexRelease {
...
+    logger.warn('hex lookup failure: No lookupName');
+    return null;
+  }
+
+  // Get dependency name from lookupName.
+  // If the dependency is private lookupName contains organization name as following:
+  // depName:organizationName
+  // depName is used to pass it in hex dep url
+  // organizationName is used for accessing to private deps
+  const depName = lookupName.split(':')[0];
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/datasource/hex/index.ts]: I found this doc, didn't try though: https://hexpm.docs.apiary.io/#introduction/authentication/api-token

[the remark this answers is not in the document]
Comment (zharinov) [reply lib/datasource/hex/index.ts]: By the way, I see the `organizationName` should go first for consistency

File: Dockerfile [hunk 28]
```diff
@@ -143,6 +168,11 @@ RUN set -ex ;\
   curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain none -y ; \
   rustup toolchain install 1.36.0
 
+# Mix and Rebar
+
+RUN mix local.hex --force
+RUN mix local.rebar --force
```
[the remark this answers is not in the document]
Comment (zharinov) [reply Dockerfile]: Yes, I think it already does (`USER ubuntu` is set little bit above, right before cargo install)

File: test/versioning/hex.spec.ts (hunk 19 above)
Comment (zharinov) [reply test/versioning/hex.spec.ts]: I've renamed it to `hexScheme` just to not confuse with actual `semver` scheme

**Comment by zharinov** [comment]
Thanks @rarkins and @ViceIce for detailed review, I've done with fixes

**Comment by zharinov** [comment]
One more thing to concern: adding notice to PR messages when lockfile update was failed, since it can cause unexpected behavior
